From b0574651cb88f00c516dc0654134a47c341e5ce7 Mon Sep 17 00:00:00 2001
From: Vincent ABRIOU <vincent.abriou@st.com>
Date: Thu, 26 Sep 2024 12:57:37 +0200
Subject: [PATCH] media: stm32: dcmipp: allow override demosaicing en via isp
 params

Allow overriding the demosaicing block enable/disabling via the
ISP param buffer.

Change-Id: I3f8d5c208f94987794c200411d475d151fcd5f75
Signed-off-by: Vincent ABRIOU <vincent.abriou@st.com>
[Alain Volmat] Commit message update
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/428455
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Alain VOLMAT <alain.volmat@st.com>
Reviewed-by: Hugues FRUCHET <hugues.fruchet@st.com>
Tested-by: Alain VOLMAT <alain.volmat@st.com>
---
 .../media/platform/st/stm32/stm32-dcmipp/dcmipp-isp-params.c  | 4 +++-
 include/uapi/linux/stm32-dcmipp-config.h                      | 2 ++
 2 files changed, 5 insertions(+), 1 deletion(-)

--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp-params.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp-params.c
@@ -288,6 +288,7 @@ dcmipp_isp_params_apply_ex(struct dcmipp
 		  cfg->shift_g << DCMIPP_P1EXCR2_SHFG_SHIFT);
 }
 
+#define DCMIPP_P1DMCR_ENABLE		BIT(0)
 #define DCMIPP_P1DMCR			0x870
 #define DCMIPP_P1DMCR_PEAK_SHIFT	16
 #define DCMIPP_P1DMCR_PEAK_MASK		0x07
@@ -321,7 +322,8 @@ dcmipp_isp_params_apply_dm(struct dcmipp
 	       DCMIPP_P1DMCR_EDGE_MASK << DCMIPP_P1DMCR_EDGE_SHIFT;
 
 	dmcr = reg_read(vout, DCMIPP_P1DMCR) & ~mask;
-	reg_write(vout, DCMIPP_P1DMCR, dmcr |
+	reg_write(vout, DCMIPP_P1DMCR, (dmcr & ~DCMIPP_P1DMCR_ENABLE) |
+		  (cfg->en ? DCMIPP_P1DMCR_ENABLE : 0) |
 		  cfg->peak << DCMIPP_P1DMCR_PEAK_SHIFT |
 		  cfg->lineh << DCMIPP_P1DMCR_LINEH_SHIFT |
 		  cfg->linev << DCMIPP_P1DMCR_LINEV_SHIFT |
--- a/include/uapi/linux/stm32-dcmipp-config.h
+++ b/include/uapi/linux/stm32-dcmipp-config.h
@@ -72,12 +72,14 @@ struct stm32_dcmipp_isp_ex_cfg {
 /**
  * struct stm32_dcmipp_isp_dm_cfg - STM32 DCMIPP ISP demosaicing filters
  *
+ * @en: enable / disable the demosaicing block
  * @edge: strength of the edge detection
  * @lineh: strength of the horizontal line detection
  * @linev: strength of the vertical line detection
  * @peak: strength of the peak detection
  */
 struct stm32_dcmipp_isp_dm_cfg {
+	__u32 en;
 	__u8 edge;
 	__u8 lineh;
 	__u8 linev;
